lib/fluent/plugin/in_tail.rb

```diff
@@ -523,6 +523,11 @@ def update_watcher(target_info, pe)
     # so adding close_io argument to avoid this problem.
     # At shutdown, IOHandler's io will be released automatically after detached the event loop
     def detach_watcher(tw, ino, close_io = true)
+      if @follow_inodes && tw.ino != ino
```

Comment:
sorry for the late response. I'll check whether can apply to !@follow_inodes condition.

On the other side, this change more like a last catch fix.
recently I'm heavily testing the fix in https://github.com/fluent/fluentd/pull/4185, which sounds a more ideal solution although it may also introduce some side-effect because the change is not so small.

From my test, it looks good under a high stress, shared some test result at https://github.com/fluent/fluentd/issues/3614#issuecomment-1575574552 :)

Reply:
I find a leak case when same log file rotates very fast, in such case, there are more than 2 watchers opened on same path with different ino, and may lead to issue:

TIME0: 
@tails[path] = ino_A // init

TIME1: 
log rotated 1st time.
Detected by ino_A on_rotate, ino_A -> ino_B. 
In update_watcher , rotated_tw = ino_A , and set @tails[path]=ino_B. detach_watcher_after_rotate_wait(rotated_tw_inoA, ino_A) 
// all looks good, ino_A can be corrected detached after wait

TIME2 (**before ino_A is detached**): 
log rotated 2nd time.
**Still** detected by **ino_A** on_rotate, ino_A -> ino_C. 
In update_watcher ,  **rotated_tw = ino_B** , and set @tails[path]=ino_C. detach_watcher_after_rotate_wait(rotated_tw_**inoB** , **ino_A**) 
// this will skip detach ino_B because the second parameter is ino_A but not expected ino_B. This happened transiently, depends on whether ino_A or ino_B on_rotate is triggered first.
// Also after this call, stop_watchers will also lose track on ino_B because it is using @tail to detect ongoing watchers, while the key is path (one path, one entry), and not it is tracking ino_C.
// Thus ino_B is leaked

Considering the solution for this leak issue:
1) change validation logic in PR: only skip detach when the rotate_tw.ino == lastest ino, so that even rotated_tw is triggerred by an even order watcher, the running rotated_tw can be detached unless it points to the latest ino (which is the previous stuck issue)
2) similar with #4185 , track @tails by ino instead of path, thus inoB won't lost track when stop_watchers.

Will continuous change the code.

Reply:
Fixed in #4208